designs/SpanMarshallers.md

```diff
@@ -0,0 +1,247 @@
+
+### Additional proposed in-source marshallers
+
+As part of this design, we would also want to include some in-box marshallers that follow the design laid out in the [Struct Marshalling design doc](./StructMarshalling.md) to support some additional scenarios:
+
+- A marshaler that marshals an empty span as `null`.
+  - This marshaller would only support empty spans as it cannot correctly represent non-empty spans of non-blittable types.
+- A marshaler that marshals out a pointer to the native memory as a Span instead of copying the data into a managed array.
+  - This marshaller would only support blittable spans by design.
+  - This marshaler will require the user to manually release the memory. Since this will be an opt-in marshaler, this scenario is already advanced and that additional requirement should be understandable to users who use this marshaler.
+  - Since there is no mechansim to provide a collection length, the question of how to provide the span's length in this case is still unresolved. One option would be to always provide a length 1 span and require the user to create a new span with the correct size, but that feels like a bad design.
+
+### Pros/Cons of Design 1
+
+Pros:
+
+- This design builds on the array support that already exists, providing implementation experience and a slightly easier implementation.
+- As we use the same MarshalAs attributes that already support arrays, developers can easily migrate their usage of array parameters in source-generated P/Invokes to use the span types with minimal hassle.
+
+Cons:
+
+- Defining custom marshalers for non-empty spans of non-blittable types generically is impossible since the marshalling rules of the element's type cannot be known.
+- Custom non-default marshalling of the span element types is impossible for non-built-in types.
+- Inlining the span marshalling fully into the stub increases on-disk IL size.
+- This design does not enable developers to easily define custom marshalling support for their own collection types, which may be desireable.
+- The MarshalAs attributes will continue to fail to work on spans used in non-source-generated DllImports, so this would be the first instance of enabling the "old" MarshalAs model on a new type in the generated DllImports, which may or may not be undesirable.
+  - The existing "native type marshalling" support cannot support marshalling collections of an unknown (at marshaller authoring time) non-blittable element type and cannot specify an element count for collections during unmarshalling.
+
+## Design 2: "Out-line" default support with extensions to Native Type Marshalling
+
+<!--
+    Idea: Extend the StructMarshalling proposal with a design for a generic collection-based marshalling pattern. That will enable outlining the Span/ReadOnlySpan/array/etc. marshallers while enabling developers to enable marshalling any collection type they want without us adding types to the generator.
+-->
```

Comment:
Examples of unmanaged APIs that would be good candidates for the generic collection-based marshaling patterns?

Reply:
Here's a few ideas I had for the patterns:

- Move arrays to use this pattern (intrinsic support in the generator)
- Span -> native array
- List -> linked-list
  - This could be useful if for example the JIT exposed a linked list type across the JIT interface body and crossgen2 needs to use it.
  - LLVM also has a few linked list types that LLVMSharp could use this with if they exposed those APIs
- Dictionary -> array of keys or Dictionary -> array of values
  - One good example of this is the coreclr hosting API
- Dictionary -> array of key-value-pair structures
  - The JVM's `JavaVMInitArgs` `options` field is a good example of this one.
- Win32 `SAFEARRAY` support could be implemented with this model for `SAFEARRAY`s containing a specific element type in the `VARIANT`s.

The nice part about this design is that one developer could provide a "marshal a Dictionary as an array of the keys" and someone else could provide a "marshal an object as the result of it's `ToString` method" and then a user who wants to pass a Dictionary down as the string values of its keys doesn't have to write a marshaller for the whole scenario. They'd only have to say "use the Dictionary keys marshaller for the dictionary, and use the `ToString` marshaller for the elements."